src/librustdoc/html/static/js/search.js

```diff
@@ -134,17 +125,220 @@ window.initSearch = function(rawSearchIndex) {
     }
 
     /**
-     * Executes the query and returns a list of results for each results tab.
-     * @param  {Object}        query          - The user query
-     * @param  {Array<string>} searchWords    - The list of search words to query against
-     * @param  {string}        [filterCrates] - Crate to search in
-     * @return {{
-     *   in_args: Array<?>,
-     *   returned: Array<?>,
-     *   others: Array<?>
-     * }}
+     * Executes the query and builds an index of results
+     *
+     * The supported syntax by this parser is as follow:
+     *
+     * ident = All characters except: " \t\n\r(<\"),>-="
+     * type-filter = [a-zA-Z]+: (one of the values in `itemTypes`).
+     * path = ident[DOUBLE-COLON ident]+
+     * generics = OPEN-ANGLE-BRACKET [arg|exact-arg],* CLOSE-ANGLE-BRACKET
+     * exact-arg = QUOTE arg QUOTE (generics)?
+     * arg = (ident|path) (generics)?
+     * args = [(arg|exact-arg)],+
+     * function-args = OPEN-PAREN [args],* END-PAREN
+     * return-args = RETURN-ARROW [args],*
+     *
+     * query = (type-filter)? (args),* (functions-args)? (return-args)?
+     *
+     * OPEN-ANGLE-BRACKET = <
+     * CLOSE-ANGLE-BRACKET = >
+     * OPEN-PAREN = (
+     * END-PAREN = )
+     * DOUBLE-COLON = ::
+     * QUOTE = "
+     * COMMA = ,
+     * RETURN-ARROW = ->
```

Comment:
I think it would be useful to specify this in actual ABNF form for clarity. For instance you are using `()` to enclose alternates. ABNF uses `/`. You are using `[]` for grouping. ABNF uses `()`.

I tried rewriting this as ABNF that passes https://tools.ietf.org/tools/bap/. I also made a few tweaks that I thought were necessary for correctness. Most notably I made `"` disallowed as part of an ident, and I restricted idents to alphanumeric (though I need to figure out how to represent Rust's / Unicode's concept of ident in ABNF)

```
ident = *1(ALPHA / DIGIT)

path = ident *WS *(DOUBLE-COLON *WS ident)
arg = path *WS [generics]
nonempty-arg-list = arg *WS *(COMMA *WS arg)
generics = OPEN-ANGLE-BRACKET *WS nonempty-arg-list *WS CLOSE-ANGLE-BRACKET
function-args = OPEN-PAREN *WS [nonempty-arg-list] *WS END-PAREN

return-args = RETURN-ARROW *WS function-args

exact-search = [type-filter] *WS QUOTE arg QUOTE
type-search = [type-filter] *WS path *WS generics
function-search = path *WS function-args *WS [return-args]

query = *WS (exact-search / type-search / function-search / return-args) *WS

type-filter = (
    "mod" /
    "externcrate" /
    "import" /
    "struct" /
    "enum" /
    "fn" /
    "type" /
    "static" /
    "trait" /
    "impl" /
    "tymethod" /
    "method" /
    "structfield" /
    "variant" /
    "macro" /
    "primitive" /
    "associatedtype" /
    "constant" /
    "associatedconstant" /
    "union" /
    "foreigntype" /
    "keyword" /
    "existential" /
    "attr" /
    "derive" /
    "traitalias")

OPEN-ANGLE-BRACKET = "<"
CLOSE-ANGLE-BRACKET = ">"
OPEN-PAREN = "("
END-PAREN = ")"
DOUBLE-COLON = "::"
QUOTE = %x22
COMMA = ","
RETURN-ARROW = "->"

ALPHA = %x41-5A / %x61-7A ; A-Z / a-z
DIGIT = %x30-39
WS = %x09 / " "
```

Reply:
> For instance you are using `()` to enclose alternates. ABNF uses `/`.

I've never seen someone use `/` for ABNF before. I think `|` is way more common (and clearer).

Reply:
[HTML uses /](https://datatracker.ietf.org/doc/html/rfc7230#appendix-B).

I've seen `|` to separate alternates before (maybe in man pages?) but AFAICT it is not valid ABNF: https://datatracker.ietf.org/doc/html/rfc5234#section-3.2

Reply:
Aha, I think `|` is the alternation character in BNF, but `/` is the alternation character in ABNF.

Reply:
I just know that in most if not all PL research papers I've read, they use `|`. _Crafting Interpreters_ also [uses](https://craftinginterpreters.com/representing-code.html#enhancing-our-notation) `|`.

Reply:
I think there are benefits from using a formal grammar. [Here's a page contrasting the two](http://xahlee.info/parser/bnf_ebnf_abnf.html):

> BNF is the original, most simple, mostly used in academic papers of theoretical context, for communicating to humans. (as opposed to being used in compiler/parser.) There's no one exact specification of BNF.
> 
> ABNF (augmented BNF) is a rather very different format that BNF, but is more standardized. It is harder to read, but is most used in parsers.
> 
> In terms of power, they are all equivalent.
>
> They are just syntactical differences. For example, in traditional BNF, the lhs/rhs separator is ::=, but in books, often →. In EBNF and ABNF it's =.

I think the fact that ABNF is standardized and has some tools available to work with it (e.g. the parser I linked) is pretty worthwhile.